pkg/controller/master/backup/backup_backing_image.go

```diff
@@ -0,0 +1,156 @@
+	return nil
+}
+
+// OnBackupBackingImageChange resync vmimage metadata files when backup backing image is completed
+func (h *backupBackingImageHandler) OnBackupBackingImageChange(_ string, backupBackingImage *lhv1beta2.BackupBackingImage) (*lhv1beta2.BackupBackingImage, error) {
+	if backupBackingImage == nil || backupBackingImage.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	if backupBackingImage.Status.State != lhv1beta2.BackupStateCompleted {
+		return nil, nil
+	}
+
+	backingImage, err := h.backingImageCache.Get(backupBackingImage.Namespace, backupBackingImage.Status.BackingImage)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, err
+	} else if backingImage == nil {
+		// if we can't find backing image, it means the backup backing image is synced from backup target
+		return nil, nil
+	}
+
+	vmImageNamespace, vmImageName := ref.Parse(backingImage.Annotations[util.AnnotationImageID])
+	if vmImageNamespace == "" || vmImageName == "" {
+		return nil, nil
+	}
+
+	vmImage, err := h.vmImageCache.Get(vmImageNamespace, vmImageName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, err
+	} else if vmImage == nil {
+		return nil, nil
+	}
+
+	target, err := settings.DecodeBackupTarget(settings.BackupTargetSet.Get())
+	if err != nil {
+		return nil, err
+	}
+
+	// when backup target is reset to default, do not trig sync
+	if target.IsDefaultBackupTarget() {
+		return nil, nil
+	}
+
+	bsDriver, err := util.GetBackupStoreDriver(h.secretCache, target)
+	if err != nil {
+		return nil, err
+	}
+
+	vmImageMetadata := &VirtualMachineImageMetadata{
+		Name:                   vmImage.Name,
+		Namespace:              vmImage.Namespace,
+		URL:                    backupBackingImage.Status.URL,
+		Checksum:               backupBackingImage.Status.Checksum,
+		DisplayName:            vmImage.Spec.DisplayName,
+		Description:            vmImage.Spec.Description,
+		StorageClassParameters: vmImage.Spec.StorageClassParameters,
+	}
+	if vmImageMetadata.Namespace == "" {
+		vmImageMetadata.Namespace = metav1.NamespaceDefault
+	}
+
+	data, err := json.Marshal(vmImageMetadata)
+	if err != nil {
+		return nil, err
+	}
+
+	shouldUpload := true
+	destPath := getVMImageMetadataFilePath(vmImage.Namespace, vmImage.Name)
+	if bsDriver.FileExists(destPath) {
+		if remoteVMImageMetadata, err := loadVMImageMetadataInBackupTarget(destPath, bsDriver); err != nil {
+			return nil, err
+		} else if reflect.DeepEqual(vmImageMetadata, remoteVMImageMetadata) {
+			shouldUpload = false
+		}
+	}
+
+	if shouldUpload {
+		logrus.Debugf("upload vm image metadata %s/%s to backup target %s", vmImage.Namespace, vmImage.Name, target.Type)
+		if err := bsDriver.Write(destPath, bytes.NewReader(data)); err != nil {
+			return nil, err
+		}
+	}
+
+	vmImageCopy := vmImage.DeepCopy()
+	harvesterv1.MetadataReady.True(vmImageCopy)
+	if !reflect.DeepEqual(vmImage, vmImageCopy) {
+		_, err = h.vmImages.Update(vmImageCopy)
+		return nil, err
+	}
```

Comment:
Hi @WebberHuang1118, thanks for the suggestion. I would prefer to put the condition in VMImage, because different Volumes may use same VMImage. In this situation, the VMImage backup metadata only need to be uploaded one time, so the condition on VMImage will be single source of truth.

Reply:
Thanks, like you said, let's keep the upload status in image.